### DISCUSSION TITLE
What is the right way to wrap lines that are longer than 80 characters?

### DISCUSSION THREAD

**Discussion by MechMel** [eclipse-langium/langium#923] [Q&A]
Is there a "right" way to detect lines that extends beyond a certain column, and wrap them?

**Comment by msujew** [comment]
Short answer: No :)

Long Answer: [It's complicated](http://journal.stuffwithstuff.com/2015/09/08/the-hardest-program-ive-ever-written/). The easiest and naive way would be just to try to look at the `$cstNode` property of any given node in the formatter and to check whether the `range.end.character-range.start.character` position is `> 80`. If it is indeed longer than 80 characters, you would need to find someway of correctly splitting it, which likely involves looking for natural splitting points such as `()[]{},.-` (or some not so natural, such as splitting a long string into multiple strings concatenated using `+`).

You can then apply the different formatting functions outlined in our [formatting guide](https://langium.org/guides/formatting/) to create newlines. The underlying algorithm usually takes good care of trying to accommodate comments in code, but it's not too sophisticated, so don't expect too much. Of course you also need to take care of the "normal" formatting as well, which might actually be in conflict with the line wrapping requirement. 

I believe it's somewhat doable actually (my initial comment recommended something else, you can take a look at the history of this answer), but still fairly tricky. It will probably require some trial and error, going up and down AST/CST nodes without creating conflicting formattings (i.e. two different formattings applied to the same part of the text).

**Reply by MechMel** [reply]
Okay, if I make a solution, I'm happy with, I'll do a write up on it.
